src/sentry/relay/config/metric_extraction.py

```diff
@@ -265,13 +285,80 @@ def _convert_widget_query_to_metric(
     return metrics_specs
 
 
+def _get_widget_cardinality_query_ttl():
+    # Add ttl + 25% jitter to query so queries aren't all made at once.
+    return int(random.uniform(_WIDGET_QUERY_CARDINALITY_TTL, _WIDGET_QUERY_CARDINALITY_TTL * 1.25))
+
+
+def _is_widget_query_low_cardinality(widget_query: DashboardWidgetQuery, project: Project):
+    """
+    Checks cardinality of existing widget queries before allowing the metric spec, so that
+    group by clauses with high-cardinality tags are not added to the on_demand metric.
+
+    New queries will be checked upon creation and not allowed at that time.
+    """
+    params: Dict[str, Any] = {
+        "statsPeriod": "1d",
+        "project_objects": [project],
+        "organization_id": project.organization_id,  # Organization id has to be specified to not violate allocation policy.
+    }
+    start, end = get_date_range_from_params(params)
+    params["start"] = start
+    params["end"] = end
+
+    query_killswitch = options.get("on_demand.max_widget_cardinality.killswitch")
+    if query_killswitch:
+        return False
+
+    if not widget_query.columns:
+        # No columns means no high-cardinality tags.
+        return True
+
+    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
+    cache_key = f"check-widget-query-cardinality:{widget_query.id}"
+    cardinality_allowed = cache.get(cache_key)
+
+    if cardinality_allowed is not None:
+        return cardinality_allowed
+
+    unique_columns = [f"count_unique({column})" for column in widget_query.columns]
+
+    query_builder = QueryBuilder(
+        dataset=Dataset.Discover,
+        params=params,
+        selected_columns=unique_columns,
+        config=QueryBuilderConfig(
+            transform_alias_to_input_format=True,
+        ),
+    )
```

Comment:
Do we know roughly how expensive this query is, and how often it will run? Any risk of hitting a timeout here?

Reply:
Yes, I'm going to be wrapping this with another exception so the config still generates, on top of that I'm going to stagger the queries by mod'ing the widget query id.